File: kirun-js/src/engine/function/system/string/PostPad.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AbstractFunction } from '../../AbstractFunction';

export class PostPad extends AbstractFunction {
    protected static readonly PARAMETER_STRING_NAME: string = 'string';

    protected static readonly PARAMETER_POSTPAD_STRING_NAME: string = 'postpadString';

    protected static readonly PARAMETER_LENGTH_NAME: string = 'length';

    protected static readonly EVENT_RESULT_NAME: string = 'result';

    protected static PARAMETER_STRING: Parameter = new Parameter(
        PostPad.PARAMETER_STRING_NAME,
        Schema.ofString(PostPad.PARAMETER_STRING_NAME),
    );

    protected static PARAMETER_POSTPAD_STRING: Parameter = new Parameter(
        PostPad.PARAMETER_POSTPAD_STRING_NAME,
        Schema.ofString(PostPad.PARAMETER_POSTPAD_STRING_NAME),
    );

    protected static PARAMETER_LENGTH: Parameter = new Parameter(
        PostPad.PARAMETER_LENGTH_NAME,
        Schema.ofInteger(PostPad.PARAMETER_LENGTH_NAME),
    );

    protected static EVENT_STRING: Event = new Event(
        Event.OUTPUT,
        new Map([[PostPad.EVENT_RESULT_NAME, Schema.ofString(PostPad.EVENT_RESULT_NAME)]]),
    );

    private signature: FunctionSignature = new FunctionSignature('PostPad')
        .setNamespace(Namespaces.STRING)
        .setParameters(
            new Map([
                [PostPad.PARAMETER_STRING.getParameterName(), PostPad.PARAMETER_STRING],
                [
                    PostPad.PARAMETER_POSTPAD_STRING.getParameterName(),
                    PostPad.PARAMETER_POSTPAD_STRING,
                ],
                [PostPad.PARAMETER_LENGTH.getParameterName(), PostPad.PARAMETER_LENGTH],
            ]),
        )
        .setEvents(new Map([[PostPad.EVENT_STRING.getName(), PostPad.EVENT_STRING]]));

    public constructor() {
        super();
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let inputString: string = context.getArguments()?.get(PostPad.PARAMETER_STRING_NAME);
        let postpadString: string = context
            ?.getArguments()
            ?.get(PostPad.PARAMETER_POSTPAD_STRING_NAME);

        let length: number = context.getArguments()?.get(PostPad.PARAMETER_LENGTH_NAME);
        let outputString: string = '';
        let prepadStringLength: number = postpadString.length;

        outputString += inputString;

        while (prepadStringLength <= length) {
            outputString += postpadString;
            prepadStringLength += postpadString.length;
        }

        if (outputString.length - inputString.length < length) {
            outputString += postpadString.substring(
                0,
                length - (outputString.length - inputString.length),
            );
        }

        return new FunctionOutput([
            EventResult.outputOf(new Map([[PostPad.EVENT_RESULT_NAME, outputString.toString()]])),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/string/post_pad.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class PostPad(AbstractFunction):

    PARAMETER_STRING_NAME: str = 'string'
    PARAMETER_POSTPAD_STRING_NAME: str = 'postpadString'
    PARAMETER_LENGTH_NAME: str = 'length'
    EVENT_RESULT_NAME: str = 'result'

    def __init__(self) -> None:
        super().__init__()

        param_string = Parameter(
            self.PARAMETER_STRING_NAME,
            Schema.of_string(self.PARAMETER_STRING_NAME),
        )
        param_postpad = Parameter(
            self.PARAMETER_POSTPAD_STRING_NAME,
            Schema.of_string(self.PARAMETER_POSTPAD_STRING_NAME),
        )
        param_length = Parameter(
            self.PARAMETER_LENGTH_NAME,
            Schema.of_integer(self.PARAMETER_LENGTH_NAME),
        )
        event_string = Event(
            Event.OUTPUT,
            {self.EVENT_RESULT_NAME: Schema.of_string(self.EVENT_RESULT_NAME)},
        )

        self._signature = (
            FunctionSignature('PostPad')
            .set_namespace(Namespaces.STRING)
            .set_parameters({
                param_string.get_parameter_name(): param_string,
                param_postpad.get_parameter_name(): param_postpad,
                param_length.get_parameter_name(): param_length,
            })
            .set_events({event_string.get_name(): event_string})
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        input_string: str = context.get_arguments().get(self.PARAMETER_STRING_NAME)
        postpad_string: str = context.get_arguments().get(self.PARAMETER_POSTPAD_STRING_NAME)
        length: int = context.get_arguments().get(self.PARAMETER_LENGTH_NAME)

        output_string = input_string
        pad_len = len(postpad_string)

        while len(output_string) - len(input_string) + pad_len <= length:
            output_string += postpad_string
            if len(output_string) - len(input_string) >= length:
                break

        current_pad = len(output_string) - len(input_string)
        if current_pad < length:
            output_string += postpad_string[:length - current_pad]

        return FunctionOutput([
            EventResult.output_of({self.EVENT_RESULT_NAME: output_string})
        ])
